contracts/VaultOUSD.sol

```diff
@@ -7,14 +7,96 @@ import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
 import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import {IERC4626} from "../contracts/interfaces/IERC4626.sol";
 import {ERC4626} from "../contracts/standard/ERC4626.sol";
+import {ParameterStore} from "./ParameterStore.sol";
+import "hardhat/console.sol";
 
 /// @title Archimedes OUSD vault
 /// @notice Vault holds OUSD managed by Archimedes under all positions.
 /// @notice It Uses ER4626 to mint shares for deposited OUSD.
 contract VaultOUSD is ERC4626 {
+    using SafeERC20 for IERC20;
+
+    ParameterStore internal _paramStore;
+    IERC20 internal _ousd;
+
+    uint256 internal _assetsHandledByArchimedes;
+
     constructor(
         IERC20Metadata asset,
         string memory name,
         string memory symbol
     ) ERC20(name, symbol) ERC4626(asset) {}
+
+    function init(address _addressParamStore, address _addressOUSD) external {
+        _paramStore = ParameterStore(_addressParamStore);
+        _assetsHandledByArchimedes = 0;
+        _ousd = IERC20(_addressOUSD);
+    }
+
+    function archimedesDeposit(uint256 assets, address receiver) external returns (uint256) {
+        _assetsHandledByArchimedes += assets;
+        return deposit(assets, receiver);
+    }
+
+    function archimedesRedeem(
+        uint256 shares,
+        address receiver,
+        address owner
+    ) external returns (uint256) {
+        uint256 redeemedAmountInAssets = redeem(shares, receiver, owner);
+        _assetsHandledByArchimedes -= redeemedAmountInAssets;
+        return redeemedAmountInAssets;
+    }
+
+    function takeRebaseFees() external {
```

Comment:
Maybe we can try to collect fee each time a user withdraws or deposit, which means we dont need an off chain element 

But users would need to pay the gas fees for this which might be a bit of an issue

Reply:
Added _takeRebaseFee to archDeposit and archRedeem, which means that ideally we dont need to monitor off chain